pkgs/development/python-modules/austin-python/default.nix

```diff
@@ -0,0 +1,88 @@
+{ stdenv
+, lib
+, buildPythonPackage
+
+, austin ? null  # Avoid circular dependency between austin / austin-python.
+, fetchFromGitHub
+, hatchling
+, hatch-vcs
+, poetry-core
+, protobuf
+, psutil
+, pytestCheckHook
+, pythonOlder
+, pythonRelaxDepsHook
+, toml
+}:
+
+buildPythonPackage rec {
+  pname = "austin-python";
+  version = "1.4.2";
+  format = "pyproject";
+
+  disabled = pythonOlder "3.7";
+
+  src = fetchFromGitHub {
+    owner = "p403n1x87";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-ddzv17lxj+59NMEe13eUNUmuzezrLY5q08+2Rrocp9E=";
+  };
+
+  postPatch = ''
+    # Fixup paths to dependencies usually found in PATH.
+    substituteInPlace austin/tools/resolve.py \
+      --replace '"addr2line"' '"${stdenv.cc.bintools}/bin/addr2line"'
+    ${if austin == null then ""
+      else ''
+        substituteInPlace austin/__init__.py \
+          --replace 'return self.BINARY' 'return "${austin}/bin/austin"'
+      ''}
```

Comment:
```suggestion
```
We can't do this in pythonPackages.

Reply:
I'm very confused: why? Clearly we can, since this PR as a whole is working fine, so you must be talking about a policy I don't know about. What is that policy?

Patching packages to find executable dependencies from the Nix store instead of relying on PATH lookup is completely standard for Nix AFAIK. Otherwise the only way to use this library will be to have `  austin` in $PATH, since the library is just a wrapper around executing the C program.

Reply:
@SuperSandro2000 what is it that can't be done?

This here makes the package free-standing, self-sufficient and hermetic PATH-wise.

A stray `austin` binary in the PATH won't override the one the package is expected to be coherent with.

In fact, [looking at the implementation](https://github.com/P403n1x87/austin-python/blob/107db3b7fd88dafc35172f1b3eb4b15b7ed74ae0/austin/__init__.py#L286), I'd go as far as arguing the whole function should be stubbed out with the hardcoded path, but I don't know enough about the use case of providing multiple methods to find the binary here. (Which seems odd anyway...)

Reply:
@samueldr I considered it but just modified the fallback instead -- the deliberate ways of pointing to another binary are hard to accidentally trigger, and unit testing relies on it too. It's a smaller diff too so all things considered... I think there are good arguments in favor of the current implementation.

Reply:
> PR as a whole is working fine

It is right now but pip is very picky about its inputs and if two of them have the same package name but other store paths you will receive an error. If this wouldn't be in place we could end up with the package being in the build inputs which would create an empty output.

This means we should reduce the amount of options for pythonPackages to a bare minimum.

Reply:
I understand the theoretical problem you're referring to, but it very obviously does not apply to this situation. There is no reason to override this parameter other than to break the circular dependency, and it's documented as such. The likelihood of two conflicting versions ending up in the same closure is pretty much 0.

The exact same thing could be done without adding options too, e.g.  `austin` depending on `austin-python.override { austin = some-dummy-derivation; }` (instead of null), it's just uglier and hackier and doesn't properly document the circular dependency problem as part of the `austin-python` derivation.